multimodal-search-tirg/TirgImageEncoder/img_text_composition_models.py

```diff
@@ -0,0 +1,256 @@
+  CVPR 2019. arXiv:1812.07119
+  """
+
+  def __init__(self, texts, embed_dim):
+    super(TIRG, self).__init__(texts, embed_dim)
+
+    self.a = torch.nn.Parameter(torch.tensor([1.0, 10.0, 1.0, 1.0]))
+    self.gated_feature_composer = torch.nn.Sequential(
+        ConCatModule(), torch.nn.BatchNorm1d(2 * embed_dim), torch.nn.ReLU(),
+        torch.nn.Linear(2 * embed_dim, embed_dim))
+    self.res_info_composer = torch.nn.Sequential(
+        ConCatModule(), torch.nn.BatchNorm1d(2 * embed_dim), torch.nn.ReLU(),
+        torch.nn.Linear(2 * embed_dim, 2 * embed_dim), torch.nn.ReLU(),
+        torch.nn.Linear(2 * embed_dim, embed_dim))
+
+  def compose_img_text(self, imgs, texts):
+    img_features = self.extract_img_feature(imgs)
+    text_features = self.extract_text_feature(texts)
+    return self.compose_img_text_features(img_features, text_features)
+
+  def compose_img_text_features(self, img_features, text_features):
+    f1 = self.gated_feature_composer((img_features, text_features))
+    f2 = self.res_info_composer((img_features, text_features))
+    f = torch.sigmoid(f1) * img_features * self.a[0] + f2 * self.a[1]
+    return f
+
+
+class TIRGLastConv(ImgEncoderTextEncoderBase):
+  """The TIGR model with spatial modification over the last conv layer.
+
+  The method is described in
+  Nam Vo, Lu Jiang, Chen Sun, Kevin Murphy, Li-Jia Li, Li Fei-Fei, James Hays.
+  "Composing Text and Image for Image Retrieval - An Empirical Odyssey"
+  CVPR 2019. arXiv:1812.07119
+  """
+
+  def __init__(self, texts, embed_dim):
+    super(TIRGLastConv, self).__init__(texts, embed_dim)
+
+    self.a = torch.nn.Parameter(torch.tensor([1.0, 10.0, 1.0, 1.0]))
+    self.mod2d = torch.nn.Sequential(
+        torch.nn.BatchNorm2d(512 + embed_dim),
+        torch.nn.Conv2d(512 + embed_dim, 512 + embed_dim, [3, 3], padding=1),
+        torch.nn.ReLU(),
+        torch.nn.Conv2d(512 + embed_dim, 512, [3, 3], padding=1),
+    )
+
+    self.mod2d_gate = torch.nn.Sequential(
+        torch.nn.BatchNorm2d(512 + embed_dim),
+        torch.nn.Conv2d(512 + embed_dim, 512 + embed_dim, [3, 3], padding=1),
+        torch.nn.ReLU(),
+        torch.nn.Conv2d(512 + embed_dim, 512, [3, 3], padding=1),
+    )
+
+  def compose_img_text(self, imgs, texts):
+    text_features = self.extract_text_feature(texts)
+
+    x = imgs
+    x = self.img_model.conv1(x)
+    x = self.img_model.bn1(x)
+    x = self.img_model.relu(x)
+    x = self.img_model.maxpool(x)
+
+    x = self.img_model.layer1(x)
+    x = self.img_model.layer2(x)
+    x = self.img_model.layer3(x)
+    x = self.img_model.layer4(x)
+
+    # mod
+    y = text_features
+    y = y.reshape((y.shape[0], y.shape[1], 1, 1)).repeat(
+        1, 1, x.shape[2], x.shape[3])
+    z = torch.cat((x, y), dim=1)
+    t = self.mod2d(z)
+    tgate = self.mod2d_gate(z)
+    x = self.a[0] * torch.sigmoid(tgate) * x + self.a[1] * t
+
+    x = self.img_model.avgpool(x)
+    x = x.view(x.size(0), -1)
+    x = self.img_model.fc(x)
+    return x
```

Comment:
can we automatically indent this file? looks wired with 2 spaces

Reply:
In the definitive version we may clone it at docker build time, so no focus on that now